source/common/router/header_formatter.cc

```diff
@@ -1,14 +1,142 @@
 #include "common/router/header_formatter.h"
 
+#include <cctype>
 #include <string>
 
 #include "common/access_log/access_log_formatter.h"
+#include "common/common/utility.h"
+#include "common/config/metadata.h"
 
 #include "fmt/format.h"
 
 namespace Envoy {
 namespace Router {
 
+namespace {
+
+// Parse list of namespace and keys into a vector. Supports backslash escaping.
+std::vector<std::string> parseUpstreamMetadataParams(const std::string& field) {
```

Comment:
Before doing a full review we should consider @htuch comment in the data-plane-api PR as to whether we potentially want to use embedded JSON here or something so we could use the JSON parser. Seems like a good idea to me to use JSON perhaps, but I could go either way. Thoughts?